File: frontend/src/services/crawlerPool.ts
import api from "./api";

export interface CookiePoolItem {
  id: string;
  label: string;
  cookie: string;
  status: "available" | "cooling" | "invalid" | "paused";
  use_count: number;
  success_count: number;
  fail_count: number;
  continuous_fail: number;
  last_used: number | null;
  last_success: number | null;
  cooling_until: number | null;
  last_error: string;
  proxy_session_id: string | null;
  proxy: {
    http: string;
    https: string;
    host?: string;
    port?: string;
    sid?: string;
  } | null;
  proxy_bound_at: number | null;
  proxy_expires_at: number | null;
  proxy_failures: number;
  created_at: string;
  updated_at: string;
}

export interface CookiePoolStats {
  total: number;
  counts: { available: number; cooling: number; invalid: number; paused: number };
  total_use: number;
  total_success: number;
  total_fail: number;
  config: {
    max_use_per_hour: number;
    max_continuous_fail: number;
    cooling_seconds: number;
    max_total_fail: number;
    proxy_session_seconds: number;
    max_proxy_failures: number;
  };
  usage_window_seconds: number;
  updated_at: string;
}

export interface ProxyPoolEntry {
  label: string;
  source: "tunnel" | "short_proxy" | "static";
}

export interface ProxyPoolStats {
  source: "tunnel" | "short_proxy" | "static" | "none";
  count: number;
  entries: ProxyPoolEntry[];
  tunnel_sids: string[];
  tunnel_period_seconds: string;
  tunnel_pool: string;
  short_proxy_refresh_seconds: number;
  static_count: number;
}

export interface CrawlCallRecord {
  ts_ms: number;
  channel: string;
  job_type: string;
  target: string;
  result: string;
  risk_type: string | null;
  error_message: string | null;
  latency_ms: number | null;
  interval_before_ms: number | null;
  proxy_used: string | null;
  cookie_id: string | null;
  cookie_label: string;
}

export const crawlerPoolService = {
  listCookies: () =>
    api.get<{ items: CookiePoolItem[]; stats: CookiePoolStats }>("/crawler/pool/cookies"),

  addCookie: (cookie: string, label: string) =>
    api.post<CookiePoolItem>("/crawler/pool/cookies", { cookie, label }),

  updateCookie: (id: string, patch: { cookie?: string; label?: string; status?: string }) =>
    api.patch<CookiePoolItem>(`/crawler/pool/cookies/${id}`, patch),

  deleteCookie: (id: string) =>
    api.delete<{ success: boolean }>(`/crawler/pool/cookies/${id}`),

  unbindCookie: (id: string) =>
    api.post<CookiePoolItem>(`/crawler/pool/cookies/${id}/unbind`),

  rebindCookie: (id: string) =>
    api.post<CookiePoolItem>(`/crawler/pool/cookies/${id}/rebind`),

  proxyStatus: () =>
    api.get<ProxyPoolStats>("/crawler/pool/proxies"),

  refreshProxies: () =>
    api.post<ProxyPoolStats>("/crawler/pool/proxies/refresh"),

  recentCalls: (limit = 50) =>
    api.get<{ items: CrawlCallRecord[] }>("/crawler/pool/calls", { params: { limit } }),
};

File: backend/app/api/v1/crawler_pool.py
"""爬虫 Cookie 池 / IP 代理池管理 API。"""
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, Field

from app.core.deps import get_current_user
from app.models.user import User

from crawler import config as crawler_config
from crawler import cookie_pool

router = APIRouter(prefix="/crawler/pool", tags=["crawl-pool"])


class CookieCreateRequest(BaseModel):
    cookie: str = Field(..., min_length=10)
    label: str = ""


class CookieUpdateRequest(BaseModel):
    cookie: str | None = None
    label: str | None = None
    status: str | None = None


@router.get("/cookies")
async def list_cookies(
    current_user: User = Depends(get_current_user),
):
    items = cookie_pool.list_cookies()
    return {"items": items, "stats": cookie_pool.pool_stats()}


@router.post("/cookies")
async def add_cookie(
    body: CookieCreateRequest,
    current_user: User = Depends(get_current_user),
):
    try:
        entry = cookie_pool.add_cookie(body.cookie, body.label)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    return entry


@router.patch("/cookies/{cookie_id}")
async def update_cookie(
    cookie_id: str,
    body: CookieUpdateRequest,
    current_user: User = Depends(get_current_user),
):
    try:
        entry = cookie_pool.update_cookie(
            cookie_id,
            label=body.label,
            cookie=body.cookie,
            status=body.status,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    if entry is None:
        raise HTTPException(status_code=404, detail="Cookie 不存在")
    return entry


@router.delete("/cookies/{cookie_id}")
async def delete_cookie(
    cookie_id: str,
    current_user: User = Depends(get_current_user),
):
    if not cookie_pool.delete_cookie(cookie_id):
        raise HTTPException(status_code=404, detail="Cookie 不存在")
    return {"success": True}


@router.post("/cookies/{cookie_id}/unbind")
async def unbind_cookie(
    cookie_id: str,
    current_user: User = Depends(get_current_user),
):
    entry = cookie_pool.unbind_cookie(cookie_id)
    if entry is None:
        raise HTTPException(status_code=404, detail="Cookie 不存在")
    return entry


@router.post("/cookies/{cookie_id}/rebind")
async def rebind_cookie(
    cookie_id: str,
    current_user: User = Depends(get_current_user),
):
    entry = cookie_pool.rebind_cookie(cookie_id, crawler_config.get_proxy_pool())
    if entry is None:
        raise HTTPException(status_code=404, detail="Cookie 不存在")
    return entry


@router.get("/calls")
async def recent_calls(
    limit: int = 50,
    current_user: User = Depends(get_current_user),
):
    limit = max(1, min(limit, 200))
    return {"items": crawler_config.recent_request_logs(limit)}


@router.post("/cookies/check")
async def check_cookies(
    user: User = Depends(get_current_user),
):
    """主动探测池内可用 Cookie 登录态（check_cookie 真实验证），返回可用性摘要。"""
    from app.services.cookie_health_scheduler import CookieHealthScheduler

    await CookieHealthScheduler()._check_job()
    cookies = cookie_pool.list_cookies()
    return {
        "total": len(cookies),
        "available": sum(1 for c in cookies if c.get("status") == "available" and c.get("cookie")),
        "cooling": sum(1 for c in cookies if c.get("status") == "cooling"),
        "invalid": sum(1 for c in cookies if c.get("status") == "invalid"),
        "has_usable": any(c.get("status") == "available" and c.get("cookie") for c in cookies),
    }


@router.get("/proxies")
async def proxy_status(
    current_user: User = Depends(get_current_user),
):
    return crawler_config.proxy_pool_stats()


@router.post("/proxies/refresh")
async def refresh_proxies(
    current_user: User = Depends(get_current_user),
):
    try:
        crawler_config.refresh_short_proxies()
    except Exception as exc:
        raise HTTPException(status_code=502, detail=f"代理刷新失败: {exc}")
    return crawler_config.proxy_pool_stats()
